### PR TITLE
add dev.k8slens.OpenLens

### PR DESCRIPTION
**Pull request by proletarius101**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:** https://github.com/lensapp/lens/issues/319#issuecomment-1385139350
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

**Comment by harmathy** [comment]
I have a similar packaging <https://github.com/harmathy/openlens-flatpak> using the debian package provided by <https://github.com/MuhammedKalkan/OpenLens>, which uses the Open Source portion of lens (thus named `dev.k8slens.openlens`). I would prefer a native (from source) build, but I haven't got very far.

**Comment by hfiguiere** [comment]
Does that mean the license in the AppStream file should say "proprietary" ?

**Comment by hfiguiere** [comment]
`innerError Error: Cannot find module '../build/Debug/pty.node'`

Why does it look for a Debug build? It should be `Release`.

File: dev.k8slens.OpenLens.yml [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        x86_64:
+          env:
+            npm_config_arch: x64
+            npm_config_target_arch: x64
+        aarch64:
+          env:
+            npm_config_arch: arm64
+            npm_config_target_arch: arm64
+    build-commands:
+      - HOME=$PWD yarn config --offline set yarn-offline-mirror $FLATPAK_BUILDER_BUILDDIR/flatpak-node/yarn-mirror
```
Comment (harmathy) [inline dev.k8slens.OpenLens.yml:64]:
```suggestion
      - HOME="$PWD" yarn config --offline set yarn-offline-mirror "$FLATPAK_BUILDER_BUILDDIR/flatpak-node/yarn-mirror"
```

File: dev.k8slens.OpenLens.yml [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+      - yarn install --cwd packages/openlens --frozen-lockfile --offline
+      - yarn --offline lerna run -- --stream --ignore open-lens build
+      - |
+        ln -srv $XDG_CACHE_HOME/node-gyp $HOME/.electron-gyp
```
Comment (harmathy) [inline dev.k8slens.OpenLens.yml:68]:
```suggestion
        ln -srv "$XDG_CACHE_HOME/node-gyp" "$HOME/.electron-gyp"
```

File: dev.k8slens.OpenLens.yml [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+      - install -Dm644 packages/open-lens/build/icons/512x512.png ${FLATPAK_DEST}/share/icons/hicolor/512x512/apps/${FLATPAK_ID}.png
+
+      # To allow separate locales
+      # https://searchfox.org/mozilla-central/rev/8a4f55bc09ffc5c25dcb4586c51ae4a9fee77b4c/taskcluster/docker/firefox-flatpak/runme.sh#131-133
+      - |
+        for lang in ${FLATPAK_DEST}/open-lens/locales/*.pak
+        do
+          locale="$(basename -s .pak $lang)"
+          install -Dm644 -t "${FLATPAK_DEST}/share/runtime/locale/${locale%%-*}/" "$lang"
+          ln -sf "${FLATPAK_DEST}/share/runtime/locale/${locale%%-*}/$(basename $lang)" "${FLATPAK_DEST}/open-lens/locales/$(basename $lang)"
```
Comment (harmathy) [inline dev.k8slens.OpenLens.yml:86]:
```suggestion
      - cp -r packages/open-lens/dist/linux-*unpacked "${FLATPAK_DEST}/open-lens"

      - install -Dm644 dev.k8slens.OpenLens.desktop "${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop"
      - install -Dm755 open-lens.sh "${FLATPAK_DEST}/bin/open-lens"
      - install -Dm644 dev.k8slens.OpenLens.metainfo.xml "${FLATPAK_DEST}/share/metainfo/${FLATPAK_ID}.metainfo.xml"
      - install -Dm644 packages/open-lens/build/icons/512x512.png "${FLATPAK_DEST}/share/icons/hicolor/512x512/apps/${FLATPAK_ID}.png"

      # To allow separate locales
      # https://searchfox.org/mozilla-central/rev/8a4f55bc09ffc5c25dcb4586c51ae4a9fee77b4c/taskcluster/docker/firefox-flatpak/runme.sh#131-133
      - |
        for lang in "${FLATPAK_DEST}"/open-lens/locales/*.pak
        do
          locale="$(basename -s .pak "$lang")"
          install -Dm644 -t "${FLATPAK_DEST}/share/runtime/locale/${locale%%-*}/" "$lang"
          ln -sf "${FLATPAK_DEST}/share/runtime/locale/${locale%%-*}/$(basename "$lang")" "${FLATPAK_DEST}/open-lens/locales/$(basename "$lang")"
```

File: open-lens.sh [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+
+set -oue pipefail
+
+export FLATPAK_ID="${FLATPAK_ID:-dev.k8slens.OpenLens}"
+export TMPDIR="${XDG_RUNTIME_DIR}/app/${FLATPAK_ID}"
+
+exec zypak-wrapper /app/open-lens/open-lens $@
```
Comment (harmathy) [inline open-lens.sh:8]:
```suggestion
exec zypak-wrapper /app/open-lens/open-lens "$@"
```

**Review by harmathy — APPROVED** [review]
Nice Job! The latest build will start up the application. 

I have a few nitpicks about quoting variables to avoid potential word splitting.

**Comment by harmathy** [comment]
I couldn't get it to connect to my local cluster. There are messages like

```
[84:0129/160823.491230:ERROR:ssl_client_socket_impl.cc(999)] handshake failed; returned -1, SSL error code 1, net_error -202
```

and 

```
warn:    ▪ [getPortFrom]: failed to retrieve port via starting to serve on (?<address>.+) +15s
```

I don't get those messages from the repackaged debian package I made, but that one is version `6.3.0`. So I have no clue if it is a packaging issue or an issue of the software itself.

**Comment by tstoeckler** [comment]
Awesome! I couldn't get this to build myself, but the prepared build works nicely :+1: 

I closed my PR now, but the one thing I had there, that isn't in this on is adding `--filesystem=~/.k8slens` to the finish args for any extensions. Not sure if that's intentional or not and I didn't actually test any extensions yet, but I had originally thought it doesn't hurt so I added it.

Also I'm still getting an error when viewing the Kubernetes configuration under _Helm Charts_:

> Error getting Helm configuration: /app/open-lens/resources/x64/helm: /app/open-lens/resources/x64/helm: File not executable.

Otherwise couldn't find anything to complain about. Thanks a lot!

**Comment by tstoeckler** [comment]
Awesome, helm is now discovered correctly :+1: thanks!

> Conventionally, app-specific config files in flatpaks are seperated from the ones in the host. Though I do need to add a --persist=~/.k8slens

Thanks for the explanation, that makes sense. Couldn't yet get the extension installation to work, but that may be an issue on my end.

**Comment by mbana** [comment]
> Build [20968 successful](https://buildbot.flathub.org/#/builders/6/builds/20968) To test this build, install it from the testing repository:
> 
> ```
> flatpak install --user https://dl.flathub.org/build-repo/3608/dev.k8slens.OpenLens.flatpakref
> ```

The install command doesn't seem to work for me:

```sh
$ flatpak install --user https://dl.flathub.org/build-repo/3608/dev.k8slens.OpenLens.flatpakref
error: Can't load uri https://dl.flathub.org/build-repo/3608/dev.k8slens.OpenLens.flatpakref: Server returned status 404
```

@tstoeckler, how did you manage to install it via `flatpak`?

**Comment by hfiguiere** [comment]
builds expire after 5 days, something like that.

**Comment by mbana** [comment]
> builds expire after 5 days, something like that.

Thanks.

What's stopping the merge then?

**Comment by mbana** [comment]
> flatpak install --user https://dl.flathub.org/build-repo/5236/dev.k8slens.OpenLens.flatpakref

@hfiguiere, I can confirm that this at least loads but fails when trying to connect:

image

## Steps to reproduce

```bash
minikube start --addons=metallb
flatpak run -v dev.k8slens.OpenLens
```

`--addons=metallb` probably does not make a difference.

## Error log

```
F0215 19:23:28.830888 172 main.go:74] failed to initialize kubeconfiginvalid configuration: [unable to read client-cert /home/mbana/.minikube/profiles/minikube/client.crt for minikube due to open /home/mbana/.minikube/profiles/minikube/client.crt: no such file or directory, unable to read client-key /home/mbana/.minikube/profiles/minikube/client.key for minikube due to open /home/mbana/.minikube/profiles/minikube/client.key: no such file or directory, unable to read certificate-authority /home/mbana/.minikube/ca.crt for minikube due to open /home/mbana/.minikube/ca.crt: no such file or directory]
goroutine 1 [running]:
k8s.io/klog/v2.stacks(0x1)
/home/runner/go/pkg/mod/k8s.io/klog/v2@v2.60.1/klog.go:860 +0x8a
k8s.io/klog/v2.(*loggingT).output(0x2df80a0, 0x3, 0x0, 0xc0001ba540, 0x1, {0x22f5723, 0x20}, 0xc000680000, 0x0)
/home/runner/go/pkg/mod/k8s.io/klog/v2@v2.60.1/klog.go:825 +0x686
k8s.io/klog/v2.(*loggingT).printDepth(0x2df80a0, 0x10018, 0x0, {0x0, 0x0}, 0x1, {0xc000490aa0, 0x2, 0x2})
/home/runner/go/pkg/mod/k8s.io/klog/v2@v2.60.1/klog.go:608 +0x1c7
k8s.io/klog/v2.(*loggingT).print(...)
/home/runner/go/pkg/mod/k8s.io/klog/v2@v2.60.1/klog.go:590
k8s.io/klog/v2.Fatal(...)
/home/runner/go/pkg/mod/k8s.io/klog/v2@v2.60.1/klog.go:1490
main.main()
/home/runner/work/lens-k8s-proxy/lens-k8s-proxy/main.go:74 +0x634

goroutine 13 [syscall]:
os/signal.signal_recv()
/opt/hostedtoolcache/go/1.17.13/x64/src/runtime/sigqueue.go:169 +0x98
os/signal.loop()
/opt/hostedtoolcache/go/1.17.13/x64/src/os/signal/signal_unix.go:24 +0x19
created by os/signal.Notify.func1.1
/opt/hostedtoolcache/go/1.17.13/x64/src/os/signal/signal.go:151 +0x2c

proxy exited with code: 255

Failed to start connection: Error: failed to retrieve port from stream
```

**Comment by harmathy** [comment]
@mbana The problem is, that access to the host filesystem for flatpak applications. For this application only the path `~/.kube/config` is accessible. 

```
flatpak override -u --filesystem="~/.minikube" dev.k8slens.OpenLens
```

This should resolve the issue for you.

**Comment by harmathy** [comment]
The same problem applies also to a custom `KUBECONFIG`. Personally I have my `KUBECONFIG` set in `~/.bash_profile`:

```
export KUBECONFIG="${XDG_DATA_HOME:-$HOME/.local/share}/kubeconfig"
```

But I guess, there is no way to programmatically expose such dynamic paths to the flatpak application.

**Comment by barthalion** [comment]
bot, build dev.k8slens.OpenLens

**Pull request closed without merging**